### ISSUE TITLE
Intermittent MDBX_BAD_DBI error

### ISSUE THREAD

**Issue by david-bouyssie** [erthink/libmdbx#276]
For still unknown reasons I sometimes get an MDBX_BAD_DBI error in my unit tests:
```The specified DBI was changed unexpectedly: MDBX_BAD_DBI: The specified DBI-handle is invalid or changed by another thread/transaction (RC=-30780)```

For instance:
https://github.com/david-bouyssie/mdbx4s/runs/5423443215?check_suite_focus=true

Do you have an idea of the underlying problem?

**Comment by david-bouyssie** [comment]
Note that this code is single thread/process.

**Comment by erthink** [comment]
Please check that you don't use a dbi-handle of kv-map which created inside the aborted transaction.
Otherwise, please show your code and a testcase.

**Comment by david-bouyssie** [comment]
The failing code is very simple:
```scala
def E_longSequential_1kBytes_withGet(): Unit = {
    
    val along = new AtomicLong()
    val random = new Random()
    for (i <- 0 until 10) {
      val tx = env.createTransaction()
      try {
        val entries = new scala.collection.mutable.HashMap[Long, Array[Byte]]
        for (j <- 0 until 100) {
          val l = along.getAndIncrement()
          val b = new Array[Byte](1024)
          random.nextBytes(b)
          db.put(tx, longToBytes(l), b)
          entries.put(l, b)
        }
        tx.commit()

        entries.foreach({ entry =>
          assertArrayEquals(db.get(longToBytes(entry._1)).get, entry._2)
        })

      } finally if (tx != null) tx.close()
    }
  }
```
10 times it:
* creates a new transaction
* puts 100 times a small array of 1024 random bytes in the current database 
* commits & closes the transaction

**Comment by erthink** [comment]
I don't know details of Castortech' bindings, but to [`put()`](https://erthink.github.io/libmdbx/group__c__crud.html#gad0e171200fd96e98de0a2d4e120fa791) a key-value pair the [dbi-handle](https://erthink.github.io/libmdbx/group__c__dbi.html#ga9bef4a9fdf27655e9343bbbf8b6fc5a1) is required.
Such handle could reference to the main key-value map (aka `MAIN_DBI`) inside a DB, or to a created named key-value map (aka subDB).

**Comment by david-bouyssie** [comment]
Actually it's not Castotrech' bindings which failed on that CI run but mine.

There is a "primary" DBI created before each test like this:
```scala
val tempDir = createTempDirectory(thisClassName, true)

val envConfig = EnvConfig(
  maxDbs = 2,
  mapSize = 16 * 1024 * 1024 // 16 MB,
)
envConfig.setFlags(MDBX_LIFORECLAIM)

env = new Env()

//Env.pushMemoryPool(1024 * 512)
env.open(tempDir, envConfig)

db = env.openDatabase("primary")
```

MDBX_LIFORECLAIM was enabled in the JUnit tests I forked from mdbxjni so I kept it.
Do you think this could explain the issue?

**Comment by erthink** [comment]
Once again, the `C` API is ```
LIBMDBX_API int mdbx_put(MDBX_txn *txn, MDBX_dbi dbi, const MDBX_val *key,MDBX_val *data, MDBX_put_flags_t flags);```.

So, where the `dbi` argument in the your code and where it comes from?

**Comment by david-bouyssie** [comment]
It comes from ```db = env.openDatabase("primary")```, which triggers mdbx_dbi_open` like this
https://github.com/david-bouyssie/mdbx4s/blob/b4845b0f21eef3655fa327a170e935c4447676b4/mdbx4s/native/src/main/scala/com/github/mdbx4s/bindings/LibraryWrapper.scala#L756

Then the DBI handle is kept as a reference of the Database class instance:
https://github.com/david-bouyssie/mdbx4s/blob/b4845b0f21eef3655fa327a170e935c4447676b4/mdbx4s/shared/src/main/scala/com/github/mdbx4s/Database.scala#L26

So the DBI handle can then be reused internally in other methods of the Database class.

**Comment by david-bouyssie** [comment]
What surprised me, is that most of the time my CI runs are ok, it fails a bit randomly.

**Comment by david-bouyssie** [comment]
As you can see it just failed on a different test:
https://github.com/david-bouyssie/mdbx4s/runs/5424926271?check_suite_focus=true

**Comment by erthink** [comment]
Seems you have got and/or used the wrong dbi-handle value.

A quick glance at your code, I found a rather [suspicious place](https://github.com/david-bouyssie/mdbx4s/blob/main/mdbx4s/jvm/src/main/scala/com/github/mdbx4s/bindings/LibraryWrapper.scala#L206).
You use the `Long` type, FAIK is 64-bit signed integer, but native `mdbx_dbi_open()` and `mdbx_dbi_open_ex()` accept a pointer to the `MDBX_dbi` which is 32-bit unsigned.
Basically this is enough for your troubles.

Suggest you should enable/use type checking (for an arguments, etc) by a compiler, and then use ASAN and/old Valrgind to verify your code.

On the other hand, I don't work with java and it's hard for me to debug your code in my head ;)
So please do it yourself ;)

**Comment by david-bouyssie** [comment]
I don't ask you to debug my code for sure ;-)

The line you mentioned correspond to the JVM implementation, which is based on the Castortheck JNI library.
The binding is indeed defined as 64-bit signed integer instead of 32-bit:
https://github.com/castortech/mdbxjni/blob/4898614740eb4070d18d5539a4111d8f2415ae35/mdbxjni/src/main/java/com/castortech/mdbxjni/JNI.java#L1176

I think this int32/int64 inconsistency might be a JNI limitation.
Also Java has not support of unsigned integer but they have implemented workarounds.

Anyway, the implementation I was referring to is the Scala Native one, which have a proper 32-bit unsigned support.
It's called UInt and used here when calling `mdbx_dbi_open`:
https://github.com/david-bouyssie/mdbx4s/blob/main/mdbx4s/native/src/main/scala/com/github/mdbx4s/bindings/LibraryWrapper.scala#L757

To be honest, I don't think the DBI is the root of this issue, since most of the time ALL the tests are working.
For instance now on Github, the two latests CI runs are green:
https://github.com/david-bouyssie/mdbx4s/actions/runs/1934747044
https://github.com/david-bouyssie/mdbx4s/actions/runs/1934763559

It means that this issue is not systematic.
This is why I'm suspecting a problem with `MDBX_LIFORECLAIM`.
I'll thus try do disable this flag to check if it increases the stability of the CI runs.

**Comment by erthink** [comment]
> This is why I'm suspecting a problem with MDBX_LIFORECLAIM.

I have a lot enough of tests going through all combinations of flags.
Ones are also succeed with Valgrind/ASAN/UBSAN.
The presence of an error can never be ruled out, but still the probability is quite small, especially in trivial scenarios.

**Comment by david-bouyssie** [comment]
I think I found & fixed the origin of the problem.
It was a garbage collection issue.
I was putting the DBI handle in an object which was collected under certain circumstances.
I changed the wrapping method to make it protected against GC:
https://github.com/david-bouyssie/mdbx4s/commit/ed459177f5831925ff2625160d3fef668ae891c1

Thank you for the discussion, it helped me to lower down the number of possibilities.
